### PR TITLE
Ghostty-quality rendering: decorations, blink, config, ligatures

### PR DESCRIPTION
**Pull request by daveowenatl**
## Summary

- **Text decorations**: Underlines (single/double/dotted/dashed/curly), strikethrough, faint/dim text, and colored underlines. Uses font-metric-based positioning from OpenType tables and anti-aliased atlas tiles for curly/dotted styles.
- **Cursor blink**: 500ms on/off cycle for blinking cursor shapes (including Default), resets on keystroke input.
- **Ghostty config loading**: New `amux-ghostty-config` crate parses `~/.config/ghostty/config` (key=value format, hex colors, palette overrides, theme references, font settings). Set `theme_source = "ghostty"` in amux config to use. Chrome colors auto-derived from terminal background.
- **Ligature support**: Run-based text shaping groups adjacent same-style cells and shapes them together through cosmic-text/HarfBuzz, enabling ligatures for programming fonts (Fira Code, JetBrains Mono, etc.). Runs break at cursor, selection, and style boundaries.
- **Bugfixes**: Ghostty backend palette override (was ignoring amux theme colors), blinking cursor shape mapping, whitespace preservation in snapshot_lines for session restore.

## Test plan

- [ ] Verify decorations: `printf '\e[4munderline\e[0m \e[9mstrike\e[0m \e[2mfaint\e[0m'` and `printf '\e[4:3mcurly\e[0m'`
- [ ] Verify cursor blink: `printf '\e[5 q'` (blinking bar), `printf '\e[0 q'` (blinking block)
- [ ] Verify Ghostty config: create `~/.config/ghostty/config` with colors, set `theme_source = "ghostty"` in amux config
- [ ] Verify ligatures: set `font_family = "Fira Code"`, type `=> -> != === <!-- -->`
- [ ] Verify session restore preserves `ls` column whitespace
- [ ] `cargo clippy --workspace -- -D warnings && cargo fmt --check && cargo test --workspace`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Load colors, palette overrides and fonts from Ghostty config files for theme/font overrides.
  * Configurable cursor blink with proper blink-on/off rendering and blink reset on input.
  * Rich text decorations: underline styles (single, double, curly, dotted, dashed), underline color, strikethrough, and faint/dim text.
  * ANSI-preserving scrollback output and extended terminal color support.

* **Chores**
  * Added a new workspace crate to parse Ghostty configs.

### CODE REVIEW

File: crates/amux-render-gpu/src/callback.rs [hunk 1]
```diff
@@ -587,6 +670,202 @@ impl CallbackTrait for TerminalPaintCallback {
...
+            if !faint_cells.is_empty() {
+                for inst in &mut fg_instances {
+                    // Map pixel position back to cell coordinates.
+                    let col = ((inst.pos[0] - self.phys_rect.x) / self.cell_width).round() as usize;
+                    let row =
+                        ((inst.pos[1] - self.phys_rect.y) / self.cell_height).round() as usize;
+                    if faint_cells.contains(&(col, row)) {
+                        inst.color[3] *= 0.5;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/callback.rs:887]: Fixed. Added bounds checking: negative positions are skipped, and col/row are clamped to valid grid range before lookup.

File: crates/amux-term/src/ghostty_pane.rs [hunk 2]
```diff
@@ -205,12 +220,32 @@ where
...
+                    } else {
+                        false
+                    };
+                    // Advance column by cell width (1 for normal, but
+                    // graphemes may contribute display width > 1).
+                    if has_content {
+                        // Already pushed chars; col tracks logical position.
+                        col = line.len();
+                    } else {
+                        col += 1;
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs]: Fixed. snapshot_lines now tracks columns using a separate char_count counter instead of line.len() (byte length), so padding and column tracking work correctly with multi-byte Unicode.

File: crates/amux-term/src/font.rs [hunk 3]
```diff
@@ -13,6 +13,18 @@ pub static SANS_SEMIBOLD: &[u8] = include_bytes!("../fonts/IBMPlexSans-SemiBold.
...
 
+/// Font decoration metrics extracted from OpenType tables (POST/OS2).
+/// All values are in physical pixels at the current font size.
+#[derive(Clone, Copy, Debug)]
+pub struct DecorationMetrics {
+    /// Distance from baseline to top of underline stroke (positive = below baseline).
+    pub underline_offset: f32,
+    /// Distance from baseline to top of strikethrough stroke (positive = above baseline).
+    pub strikeout_offset: f32,
+    /// Recommended stroke thickness for underline/strikethrough.
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/font.rs]: Fixed. Updated docs to state values are in logical points (not physical pixels), and that callers multiply by pixels_per_point.

File: crates/amux-term/src/font.rs [hunk 4]
```diff
@@ -102,3 +114,63 @@ pub fn measure_cell_width(font_system: &mut FontSystem, metrics: Metrics) -> f32
...
+    font_size: f32,
+) -> DecorationMetrics {
+    use cosmic_text::ttf_parser;
+
+    // Find the first monospace face in the font database.
+    let mono_id = font_system
+        .db()
+        .faces()
+        .find(|f| f.monospaced)
+        .map(|f| f.id);
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/font.rs]: Fixed. Now uses fontdb::Query with Family::Monospace to resolve the configured monospace face, honoring set_monospace_family().

File: crates/amux-render-gpu/src/callback.rs [hunk 5]
```diff
@@ -587,6 +670,202 @@ impl CallbackTrait for TerminalPaintCallback {
...
+            let line_thickness = (dm.stroke_size * pixels_per_point).max(1.0);
+            // Underline: below baseline. The font's underline_offset is distance
+            // from baseline (positive = below). Baseline is at ~line_top + ascent.
+            // For a terminal cell: baseline is approximately at line_y from cosmic-text.
+            // We use the font's own metric scaled by ppem.
+            let baseline_y = self.cell_height * 0.78; // approximate baseline
+            let underline_y_offset = (baseline_y + dm.underline_offset * pixels_per_point).max(0.0);
+            let strikethrough_y_offset =
+                (baseline_y - dm.strikeout_offset * pixels_per_point - line_thickness / 2.0)
+                    .max(0.0);
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/callback.rs:704]: Low severity — deferring. The 0.78 ratio works well across common monospace fonts and matches the cosmic-text baseline behavior for terminal rendering. Deriving from per-row layout runs would add complexity for marginal benefit.

**Comment by daveowenatl** [comment]
@coderabbitai review

File: crates/amux-term/src/ghostty_pane.rs [hunk 6]
```diff
@@ -207,12 +222,33 @@ where
...
+                            true
                         }
+                    } else {
+                        false
+                    };
+                    // Advance column by 1 per cell.
+                    if has_content {
+                        col = char_count;
+                    } else {
+                        col += 1;
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs]: Fixed. snapshot_lines now uses cell index (col += 1 per cell) with a separate visible_col counter, matching snapshot_lines_ansi's approach.

File: crates/amux-term/src/ghostty_pane.rs [hunk 7]
```diff
@@ -221,6 +257,128 @@ where
...
+                        }
+                        if cur_style.faint {
+                            line.push_str(";2");
+                        }
+                        if cur_style.italic {
+                            line.push_str(";3");
+                        }
+                        match cur_style.underline {
+                            GhosttyUnderline::Single => line.push_str(";4"),
+                            GhosttyUnderline::Double => line.push_str(";21"),
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs:322]: Fixed. Now emits curly (4:3), dotted (4:4), dashed (4:5) extended underline SGR, plus underline color via SGR 58;2;R;G;B when present.

File: crates/amux-term/src/ghostty_pane.rs [hunk 8]
```diff
@@ -139,36 +142,48 @@ where
...
-                    selection_bg: Color(0.4, 0.6, 1.0, 1.0),
-                    colors: colors.palette.iter().map(|&c| rgb_to_color(c)).collect(),
-                };
+            // Cache palette (skip if externally overridden by amux theme).
+            if !self.palette_overridden {
+                if let Ok(colors) = snapshot.colors() {
+                    let fg = rgb_to_color(colors.foreground);
+                    let bg = rgb_to_color(colors.background);
+                    let cursor_color = colors.cursor.map(rgb_to_color).unwrap_or(fg);
+                    self.cached_palette = Palette {
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs:170]: Valid concern — deferring for now. libghostty-vt 0.1.x doesn't expose runtime palette change events, so there's no signal to clear the override. When we upgrade to a version with palette change callbacks, we can implement the merge approach.

File: crates/amux-render-gpu/src/callback.rs [hunk 9]
```diff
@@ -879,8 +1163,165 @@ impl CallbackTrait for TerminalPaintCallback {
...
+    cell_height: f32,
+    phys_x: f32,
+    phys_y: f32,
+    pixels_per_point: f32,
+    resources: &mut TerminalGpuResources,
+    queue: &wgpu::Queue,
+    fg_instances: &mut Vec<CellFgInstance>,
+    cached_glyphs: &mut Vec<CachedGlyph>,
+) {
+    let cache_key = (run.text.clone(), run.bold, run.italic);
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/callback.rs:1202]: Fixed. Added last_pixels_per_point tracking to TerminalGpuResources — shape_cache, curly_tile, and dotted_tile are all cleared when DPI changes.

File: crates/amux-render-gpu/src/callback.rs [hunk 10]
```diff
@@ -587,6 +670,207 @@ impl CallbackTrait for TerminalPaintCallback {
...
+                    resources.atlas_bind_group_dirty = true;
+                }
+            }
+            if resources.dotted_tile.is_none() {
+                let (w, h, data) = rasterize_dotted_tile(self.cell_width, line_thickness);
+                if let Some(entry) = resources.atlas.insert_raw_mono(queue, w, h, &data) {
+                    resources.dotted_tile = Some(entry);
+                    resources.atlas_bind_group_dirty = true;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/callback.rs:720]: Fixed. Covered by the same DPI invalidation — curly_tile and dotted_tile are reset to None when pixels_per_point changes, forcing re-rasterization.

File: crates/amux-app/src/render.rs [hunk 11]
```diff
@@ -53,6 +54,11 @@ pub(crate) fn render_pane(
             current_highlight,
         );
 
+        // Apply cursor blink: hide cursor during "off" phase for blinking shapes.
+        if !cursor_blink_on && is_blinking_cursor(snapshot.cursor_shape) {
+            snapshot.cursor_visible = false;
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/render.rs:62]: Fixed. Added cursor_blink_hidden field to TerminalSnapshot. Blink state no longer touches cursor_visible, so ligature run-breaking stays stable across blink cycles.

File: crates/amux-app/src/theme.rs [hunk 12]
```diff
@@ -95,6 +95,78 @@ impl Theme {
...
+        if let Some(c) = cfg.selection_foreground() {
+            terminal.selection_fg = c;
+        }
+
+        // Apply ANSI palette overrides (0-15).
+        for i in 0..16u8 {
+            if let Some(c) = cfg.ansi_color(i) {
+                terminal.ansi[i as usize] = c;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/theme.rs]: Fixed. Theme now stores palette_overrides HashMap and apply_to_palette propagates all 0-255 palette entries from Ghostty config.

File: crates/amux-ghostty-config/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+                continue;
+            }
+
+            // Split on first '=' only.
+            let Some((key, value)) = line.split_once('=') else {
+                continue;
+            };
+            let key = key.trim();
+            let value = value.trim();
+
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-ghostty-config/src/lib.rs:92]: Fixed. config_search_paths() and load_theme() now use dirs::config_dir() which respects XDG_CONFIG_HOME.

File: crates/amux-ghostty-config/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+/// Standard config file search paths.
+fn config_search_paths() -> Vec<PathBuf> {
+    let mut paths = Vec::new();
+
+    // XDG / Linux / cross-platform: ~/.config/ghostty/config
+    if let Some(home) = dirs::home_dir() {
+        paths.push(home.join(".config").join("ghostty").join("config"));
+    }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-ghostty-config/src/lib.rs]: Fixed. Parser now strips matching surrounding single or double quotes from values before storing.

File: crates/amux-render-gpu/src/atlas.rs [hunk 15]
```diff
@@ -279,4 +279,32 @@ impl GlyphAtlas {
...
+                (x + width) as f32 / s,
+                (y + height) as f32 / s,
+            ],
+            placement_left: 0,
+            placement_top: 0,
+            width,
+            height,
+            is_color: false,
+        })
+    }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/atlas.rs:312]: Fixed. insert_raw_mono now returns None early if width or height is 0.

File: crates/amux-render-gpu/src/callback.rs [hunk 16]
```diff
@@ -587,6 +670,207 @@ impl CallbackTrait for TerminalPaintCallback {
...
+            let curly_tile = resources.curly_tile;
+            let dotted_tile = resources.dotted_tile;
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-render-gpu/src/callback.rs:722]: Fixed. Added last_pixels_per_point to TerminalGpuResources — shape_cache, curly_tile, and dotted_tile are all invalidated when DPI changes.

File: crates/amux-term/src/ghostty_pane.rs [hunk 17]
```diff
@@ -207,12 +222,33 @@ where
...
                         }
+                    } else {
+                        false
+                    };
+                    // Advance column by 1 per cell.
+                    if has_content {
+                        col = char_count;
+                    } else {
+                        col += 1;
                     }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs]: Fixed. Both snapshot_lines and snapshot_lines_ansi now use cell index (col += 1 per cell) with a separate visible_col counter.

File: crates/amux-term/src/ghostty_pane.rs [hunk 18]
```diff
@@ -221,6 +257,128 @@ where
...
+                        if let Some(fg) = cur_fg {
+                            use std::fmt::Write;
+                            let _ = write!(line, ";38;2;{};{};{}", fg.r, fg.g, fg.b);
+                        }
+
+                        // Background color
+                        if let Some(bg) = cur_bg {
+                            use std::fmt::Write;
+                            let _ = write!(line, ";48;2;{};{};{}", bg.r, bg.g, bg.b);
+                        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/ghostty_pane.rs:365]: Fixed. snapshot_lines_ansi now emits curly (4:3), dotted (4:4), dashed (4:5) SGR and underline color (58;2;R;G;B).

File: crates/amux-term/src/pane.rs [hunk 19]
```diff
@@ -595,14 +596,35 @@ impl TerminalPane {
...
                     text: cell_ref.str().to_string(),
                     fg: Color::from(fg_srgba),
                     bg: Color::from(bg_srgba),
                     bold: attrs.intensity() == wezterm_term::Intensity::Bold,
                     italic: attrs.italic(),
-                    underline: attrs.underline() != wezterm_term::Underline::None,
+                    underline,
+                    underline_color,
                     strikethrough: attrs.strikethrough(),
+                    faint: attrs.intensity() == wezterm_term::Intensity::Half,
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-term/src/pane.rs:626]: Fixed. Removed the extra blank line.

**Comment by daveowenatl** [comment]
## Addressing review body findings (round 2)

### CodeRabbit — Outside diff range
- **Ligature selection/search recoloring** (callback.rs:492-609): Valid observation that partial selection of a ligature glyph won't recolor correctly on the appearance-only path. Deferring — this is a known limitation of the run-cache approach. A proper fix requires splitting/re-shaping ligatures at selection boundaries, which is significant work for an edge case.

### CodeRabbit — Nitpick
- **Theme merge precedence and theme-file discovery tests** (ghostty-config/lib.rs:267-312): Good suggestion. Will add test coverage for load_from theme merging and load_theme path discovery in a follow-up.

### Copilot round 1 (already addressed in fd9ef7c)
- palette_overridden concern: Deferred (libghostty-vt 0.1.x lacks runtime palette change events)
- baseline_y magic constant: Deferred (works well in practice, low ROI to change)
